### ISSUE TITLE
Consider removing GPUPipelineLayout

### ISSUE THREAD

**Issue by kainino0x** [gpuweb/gpuweb#2550]
The motivation for this is:

- Partially implicit pipeline layouts: https://github.com/gpuweb/gpuweb/issues/2470#issuecomment-1024831910
- Pipeline layouts are entirely represented by their value.
- The driver *probably* doesn't do much work inside the creation of a pipeline layout objects.
- WebGPU doesn't actually take advantage of pipeline layout objects. There's no pipeline layout object in setBindGroup (like in vkCmdBindDescriptorSets).

Editors weren't 100% sure if we still might want pipeline layout objects in the future (for things like push constants, immutable samplers, pipeline layout compatibility?), though they could always be added back.

Sorry, going on vacation now. More details later :P

**Comment by Kangz** [comment]
IMHO since pipeline layouts are compared by value-equality, we don't need the objects, just the descriptor. I can't find it but it was suggested to remove `GPUPipelineLayout` at least twice.

Now, does this change clear the bar for v1? Personally I'm leaning yes since it is a clear simplification that unlocks #2470 that's a good ergonomic change, but doesn't structurally change WebGPU.

**Comment by Kangz** [comment]
Putting the V1 tag so we can discuss whether it should be in V1 or not.

**Comment by kvark** [comment]
I've thought about this more than necessary. On one hand, it corresponds to a real API object in both Vulkan and D3D12.

On the other hand, in Vulkan at least the `VkPipelineLayout` is separate because it's provided in a bunch of places, like `vkCmdBindDescriptorSets`. This isn't the case for WebGPU (which in turns means our pipeline layout isn't *really* the same semantics as Vulkan's). In WebGPU we'd only be specifying the pipeline layout at pipeline creation, which by itself can be relatively heavy already.

TL;DR: I think it's fine to do, and we should probably do it.

**Comment by litherum** [comment]
Sounds like we need a perf investigation. I'm probably not the right one to perform this investigation given it isn't relevant on Metal.

**Comment by Kangz** [comment]
This is a cacheable thing that's done at pipeline compilation time. The CPU overhead will be very minimal, and the GPU commands produced the same.

**Comment by kdashg** [comment]
<details><summary><a href="https://github.com/gpuweb/gpuweb/wiki/gpu-web-2022-02-16#consider-removing-gpupipelinelayout-2550">WebGPU meeting minutes 2022-02-16</a></summary>



* KN: in editor's meeting - concerns about getBindGroupLayout. (1) rather than automatic pipeline layouts, automatic bind group layouts. At that point GPUPipelineLayout object becomes pointless. These are orthogonal. Can do one without the other.
* GPUPipelineLayouts - value objects. No internal state. Typically represent an object in the API. We can do value-to-API-object cache keyed on (…)
* It's an object in the API used to pre-bake a value. Not super necessary.
* BGLs are similar, but used more widely. Passed to pipelines, bind groups. Pipeline layouts mostly only passed in pipeline creation, can be passed by value there.
* MM: what's the benefit?
* KN: getting rid of boilerplate of creating pipeline layout so you can immediately use it. Can do things with pipeline layout description - e.g. partial. Spec one for one BGL, leave one undefined, let default algorithm to it for me. Make automatically gen'd pipeline layouts finer grained.
* MM: could keep pipeline object and say if you want to create one of these you have to fully fill it out. Could supply either it, or array of bind groups.
* KN: yes. That's why they're orthogonal. Original idea was to do that. Pass array of bind groups rather than pipeline layout object.
* MM: fine with me.
* KG: sounds like consensus is we should try this.
* MM: sure.
* KN: primary concern - making changes at this date where arguably not necessary. This is definitely not necessary.
* BJ: implicit pipeline layouts have been consistent source of developer complaint. Confusing. We need better communicated errors. All-or-nothing thing right now. This would make them much more useful. Could share same frame/camera uniforms across all pipelines, and have pipeline only implicitly fill out material variables they need. Not necessary, but useful change addressing most complaints about API I've heard.
* KG: needs-spec then.
* KN: even if we decide we want to put off finer-grained implicit layouts - this makes it cleaner to add that later. That is a bigger change. This one is smaller.
* MM: point of implicit layouts was to make devs' lives easier. If they're adding complexity we have to do something here.
* BJ: they're useful in certain narrow scenarios now. This will make them more generally useful.
* MM: OK.


</details>

**Comment by kainino0x** [comment]
Editors (Kai/Brandon) not feeling this is worth the effort/changes, propose closing.

**Comment by kainino0x** [comment]
Closing unless someone wants to reopen.

**Comment by kdashg** [comment]
[GPU Web 2022-03-02 (APAC)](https://docs.google.com/document/d/1TbMOUMFNJgW_7M0YvYfuiH4m66V3xyuvDwEvmR36vWQ):
* MM: Does GPUPipelineLayout correspond to anything in D3D?
* AE: It sort of does to the root signature, but only imperfectly.
* MM: But there’s other stuff in the RS, yeah? (yes, thus imperfectly)
* MM: We probably shouldn’t have this bake step if it’s generally unused by native APIs.
* AE: We do use it, but I don’t know off the top of my head why/for what.
* BJ: It’s not that we think we should keep it, but rather we don’t think it’s worth changing this late in the process, and I think it’s a fairly minimal wart.

**Comment by kvark** [comment]
> AE: It sort of does to the root signature, but only imperfectly.

Not sure what this is referring to. In wgpu the relation is exactly 1:1 with root signatures, so it's sort of a perfect mapping for us.

**Comment by austinEng** [comment]
> > AE: It sort of does to the root signature, but only imperfectly.
> 
> Not sure what this is referring to. In wgpu the relation is exactly 1:1 with root signatures, so it's sort of a perfect mapping for us.

No concerns from me; we are in alignment.
During the meeting, I was pretty sure we have exactly one root signature per pipeline layout in the D3D12 backend. I wasn't 100% certain without checking the code, so what I said was more along the lines of "we have exactly one root signature per pipeline layout; if the mapping isn't perfect, I couldn't tell you off the top of my head what we do to handle the difference".
